platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java

```diff
@@ -65,42 +65,46 @@ private DefaultProblemBuilder createProblemBuilder() {
     }
 
     @Override
-    public RuntimeException throwing(Action<ProblemSpec> spec) {
+    public RuntimeException throwing(Throwable exception, Action<? super ProblemSpec> spec) {
         DefaultProblemBuilder problemBuilder = createProblemBuilder();
         spec.execute(problemBuilder);
+        problemBuilder.withException(exception);
+        report(problemBuilder.build());
+        throw runtimeException(exception);
+    }
 
-        Problem problem = problemBuilder.build();
-        Throwable exception = problem.getException();
-        if (exception == null) {
-            throw new IllegalStateException("Exception must be non-null");
-        } else {
-            throw throwError(exception, problem);
-        }
+    @Override
+    public RuntimeException throwing(Throwable exception, Problem problem) {
+        problem = ((InternalProblem) problem).toBuilder(additionalDataBuilderFactory).withException(transform(exception)).build();
+        report(problem);
+        throw runtimeException(exception);
     }
 
     @Override
     public RuntimeException throwing(Throwable exception, Collection<? extends Problem> problems) {
         for (Problem problem : problems) {
             report(((InternalProblem) problem).toBuilder(additionalDataBuilderFactory).withException(transform(exception)).build());
         }
-        if (exception instanceof RuntimeException) {
-            return (RuntimeException) exception;
-        } else {
-            throw new RuntimeException(exception);
-        }
+        throw runtimeException(exception);
     }
 
-    private RuntimeException throwError(Throwable exception, Problem problem) {
-        report(problem);
+    private static RuntimeException runtimeException(Throwable exception) {
```

Comment:
💄We might want to denote to ourselves that this wraps something? E.g. `wrapsIntoRuntimeException`

Reply:
Or just `wrapInto`, the return type and parameter should make it also quite trivial.

Reply:
We don't always wrap the exception so `maybeWrapInto` seems to be the most fitting name